webui/react/src/utils/task.ts

```diff
@@ -88,3 +90,42 @@ export const canBeOpened = (task: AnyTask): boolean => {
   if (isExperiment(task)) return true;
   return !!task.url;
 };
+
+const matchSearch = <T extends AnyTask>(task: T, search = ''): boolean => {
+  if (!search) return true;
+  return task.id.indexOf(search) !== -1 || task.title.indexOf(search) !== -1;
+};
+
+const matchesState = <T extends AnyTask>(task: T, states: string[]): boolean => {
+  if (states[0] === ALL_VALUE) return true;
+
+  const targetStateRun = states[0] as RunState;
+  const targetStateCmd = states[0] as CommandState;
+
+  return [ targetStateRun, targetStateCmd ].includes(task.state);
+};
+
+const matchesUser = <T extends AnyTask>(task: T, users: User[], username?: string): boolean => {
+  if (!username) return true;
+  const selectedUser = users.find(u => u.username === username);
+  return !!selectedUser && (task.ownerId === selectedUser.id);
+};
+
+export const filterTasks = <T extends TaskType = TaskType, A extends AnyTask = AnyTask>(
+  tasks: A[], filters: TaskFilters<T>, users: User[], search = '',
+): A[] => {
+  const isAllTypes = !Object.values(filters.types).includes(true);
+  return tasks
```

Comment:
question: Not sure I'm reading this correctly, but are we doing a loop for each call to `.filter`? Would be more efficient as a single `.filter` call, but more annoying to write for sure. Given that this list can get sizable it might be worth the tradeoff to save a few multiples of `n`, but I'll let you make that call as it's not a full order of magnitude difference.

Reply:
That's a good point. It will loop through the output of the previously filtered list at each `.filter` call. Condensing the filters into a single call.